cpp/src/arrow/python/builtin_convert.cc

```diff
@@ -157,8 +179,15 @@ class ScalarVisitor {
   int64_t float_count_;
   int64_t binary_count_;
   int64_t unicode_count_;
+  int64_t decimal_count_;
+
+  int32_t max_decimal_precision_;
+  int32_t max_decimal_scale_;
+
   // Place to accumulate errors
   // std::vector<Status> errors_;
+  OwnedRefNoGIL decimal_module_;
```

Comment:
Is this actually required? You are only using the decimal type below.

Reply:
I might be misunderstanding the use case for `OwnedRefNoGIL`. 

My understanding is that if we're `Py_XDECREF`ing something and we do not hold then GIL then we need to acquire it.

The GIL is released in Cython before the function that instantiates this class is called (and isn't subsequently acquired in that function). Assuming my understanding is correct then we need to acquire the GIL before calling `Py_XDECREF` on this.

Are you suggesting that because the order of destruction here is guaranteed to be the reverse of initialization order that we don't need to care that `decimal_module_` holds the GIL because we'll never decref `decimal_module_` before we decref `decimal_type_`?

Reply:
What I mean is that you don't care about keeping a reference to the decimal module if you only need to use the decimal type:
```
>>> D = __import__('decimal').Decimal
>>> sys.modules['decimal'] = None
>>> D(100)
Decimal('100')
```

Reply:
Oh, awesome. There are a few places that we have unnecessary refs to the `decimal` module. I'll open a JIRA to clean those up (and remove this one). Thanks for the review.